packages/eslint-config-react-app/index.js

```diff
@@ -10,8 +10,6 @@
 // Inspired by https://github.com/airbnb/javascript but less opinionated.
 
 // We use eslint-loader so even warnings are very visible.
-// This is why we only use "WARNING" level for potential errors,
```

Comment:
Instead of removing it would be better to clarify the criteria we use to decide which it should be

Reply:
Yeah, fair enough. Umm...So from what I understand, we have some eslint rules that are configured as errors to catch run time errors without having to build and run the app.

 - 'react/jsx-no-undef'
 - 'react/react-in-jsx-scope'
 - 'react/require-render-return'

Makes sense that these are errors.

We also have some eslint rules that are configured as errors to prevent users from using features that aren't supported by CRA:

 - 'import/first'
 - 'import/no-amd'
 - 'import/no-webpack-loader-syntax'
 - 'no-restricted-properties'

Makes sense that these are errors too.

But, the following three eslint rules that are configured as errors are a bit different. If I understand correctly these are in place to help educate or train CRA users to follow certain best practices. But couldn't the same be said for every other eslint rule we have? I guess what I'm wondering, is how can we justify that these are set as errors while other eslint rules are set as warnings. Is it the severity of not following these practices? Or the difficultly of tracking bugs that could arise from not following these practices?

 - 'import/first'
 - 'no-restricted-globals'
 - 'no-undef'

Reply:
`no-undef` is a runtime error 99% of the time, the other 1% of the time doesn't make sense in a bundled system anyway (referencing `$` instead of `window.$` when using a script tag). To our build tool's knowledge, it will be a build error until the user alters the environment at runtime.

Reply:
`no-restricted-globals` is a tricky one, because it serves a few purposes -- first, many people are unaware of the entire list of globals, causing very confusing errors, e.g.:
```js
handleClick() {
  // other preface code
  event.target.value; // why is this blowing up ?!
}
```

Secondly, we want to keep a future of SSR applications open -- training people to use the `window.` prefix now will make this transition easier when running in a Node environment.

Reply:
> no-undef is a runtime error 99% of the time
Yeah sorry, I should have included this with the other run-time error eslint rules we have.

Reply:
`import/first` is arguably a style choice -- but it bites our users so often that I think an exclusion can be made here.

People try stuff like this all the time, and it looks like it would work:
```js
import $ from 'jquery';
window.jQuery = $;
import 'jquery-slimscroll'; // this module relies on jQuery being global
```

When in reality, that code is actually evaluated as:
```js
import $ from 'jquery';
import 'jquery-slimscroll';
window.jQuery = $;
```

In this case, you can argue it's not a style choice because they're not semantically identical.

Reply:
> If I understand correctly these are in place to help educate or train CRA users to follow certain best practices. But couldn't the same be said for every other eslint rule we have? I guess what I'm wondering, is how can we justify that these are set as errors while other eslint rules are set as warnings. Is it the severity of not following these practices? Or the difficultly of tracking bugs that could arise from not following these practices?

Hopefully my three explanations helped you understand our position a bit more. 😄 
I don't think other rules have such a strong argument.

Reply:
> Hopefully my three explanations helped you understand our position a bit more. :smile:
I don't think other rules have such a strong argument.

Kinda? Haha. Your explanations were helpful in understanding why `import/first` and `no-restricted-globals` are necessary, and why its a good idea to have them set as errors.

I guess what I'm still iffy on is the _criteria_ we're using to decide whether an eslint rule should be configured as a warning or error.

There are two criteria for making a rule an error that are crystal clear to me:
 
1. Run-time errors
2. Features that aren't supported
3. ???

What you said here is more what I was looking for:

> but it bites our users so often that I think an exclusion can be made here.

1. Run-time errors
2. Features that aren't supported
3. _stuff that causes our users so much grief that it's better that these are errors_?

Know what I mean? Sorry for being a pain in the ass. I hope y'all don't view this discussion as pedantic and unnecessary. It may help to have these criteria laid out though. In case others ask in the future.

Reply:
You're not a pain in the ass at all. 😄

At the end of the day, we do want to enforce what's best for majority of users.
So I feel like "stuff that causes our users so much grief that it's better that these are errors" is a pretty accurate description.

Perhaps more accurately, error for anything which has a non-obvious/indeterministic behavior to the majority of our users.

The `import` case requires a deep understanding of the spec and language features -- a lot of the community doesn't have this understanding yet (I even trip up on it occasionally).
Babel blurs this functionality, so it's probably going to remain one of those things we need to enforce even though advanced users know the caveats.

Maybe in a few years when `import`/`export` are mainstream in Node & browsers, and no longer require Babel we can remove the `import/first` rule.
But right now, it's so confusing because existing implementations don't follow spec / etc.